src/Catalog.API/Services/CatalogAI.cs

```diff
@@ -1,55 +1,76 @@
-﻿using Azure.AI.OpenAI;
+﻿using System.Diagnostics;
+using Microsoft.SemanticKernel.Embeddings;
 using Pgvector;
 
 namespace eShop.Catalog.API.Services;
 
 public sealed class CatalogAI : ICatalogAI
 {
-    private readonly string _aiEmbeddingModel;
-    private readonly OpenAIClient _openAIClient;
+    private const int EmbeddingDimensions = 384;
+    private readonly ITextEmbeddingGenerationService _embeddingGenerator;
 
     /// <summary>The web host environment.</summary>
     private readonly IWebHostEnvironment _environment;
     /// <summary>Logger for use in AI operations.</summary>
     private readonly ILogger _logger;
 
-    public CatalogAI(IOptions<AIOptions> options, IWebHostEnvironment environment, ILogger<CatalogAI> logger, OpenAIClient openAIClient = null)
+    public CatalogAI(IWebHostEnvironment environment, ILogger<CatalogAI> logger, ITextEmbeddingGenerationService embeddingGenerator = null)
     {
-        var aiOptions = options.Value;
-        _openAIClient = openAIClient;
-        _aiEmbeddingModel = aiOptions.OpenAI.EmbeddingName ?? "text-embedding-ada-002";
-        IsEnabled = _openAIClient != null;
+        _embeddingGenerator = embeddingGenerator;
         _environment = environment;
         _logger = logger;
+    }
+
+    /// <summary>Gets whether the AI system is enabled.</summary>
+    public bool IsEnabled => _embeddingGenerator is not null;
+
+    /// <summary>Gets an embedding vector for the specified catalog item.</summary>
+    public ValueTask<Vector> GetEmbeddingAsync(CatalogItem item) =>
+        IsEnabled ?
+            GetEmbeddingAsync(CatalogItemToString(item)) :
+            ValueTask.FromResult<Vector>(null);
 
-        if (_logger.IsEnabled(LogLevel.Information))
+    /// <summary>Gets embedding vectors for the specified catalog items.</summary>
+    public async ValueTask<IReadOnlyList<Vector>> GetEmbeddingsAsync(IEnumerable<CatalogItem> items)
+    {
+        if (IsEnabled)
         {
-            _logger.LogInformation("Embedding model: \"{model}\"", _aiEmbeddingModel);
+            long timestamp = Stopwatch.GetTimestamp();
+
+            IList<ReadOnlyMemory<float>> embeddings = await _embeddingGenerator.GenerateEmbeddingsAsync(items.Select(CatalogItemToString).ToList());
+            var results = embeddings.Select(m => new Vector(m[0..EmbeddingDimensions])).ToList();
```

Comment:
Just for my knowledge, is there no way to pass `EmbeddingDimensions` into the embedding generator? That way it doesn't need to do extra work just to get discarded later.